### PR TITLE
More type safe implementation of BType, cleanups in GenBCode

### PR DESCRIPTION
**Pull request by lrytz**
More type safe `BType` implementation. Various cleanups in the `GenBCode` backend. Also some cleanups and clarifications in `Names`. More details in the individual commit messages.

### CODE REVIEW

**Comment by lrytz** [comment]
Review by @gkossakowski (everybody's feedback is appreciated, of course).

**Comment by lrytz** [comment]
So one problem we have is that PR validation doesn't run with `-Ybackend:GenBCode`, we only have a nightly. I'm running a [scala-checkin-manual](https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/1218/parameters/) with the appropriate compiler flags. Is there an urge to automate that? I'm OK to do it manually for my PRs, and that might be enough since the code is under `-Y`.

**Comment by gkossakowski** [comment]
I think it's fine to use scala-checkin-manual for now. However, we shouldn't have any failures with regular pr validation right?

**Comment by lrytz** [comment]
So the first of the two is expected, it's explained in the commit message. It simplifies reviewing, I'll squash it after (I can also do it now, as you prefer). The second should have built successfully, I'll look into it.

I'll also adding some unit tests for `reflect.internal.Names`.

**Comment by gkossakowski** [comment]
Ah, I didn't look at commit messages. I'll start reviewing now.

**Comment by lrytz** [comment]
@gkossakowski I fixed the build issue and squashed the refactoring so that all commits build. If you want to review https://github.com/lrytz/scala/commit/70d179084673fa317c4380529c9ab58f42de19e8 in two steps, they are https://github.com/lrytz/scala/commit/476d837a9fca33290b7f3a374c18d9ccab5471ba and https://github.com/lrytz/scala/commit/d185b647a5fd4d4ca15712f1bfdfb57497bef1a2.

**Comment by adriaanm** [comment]
PLS SYNCH

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala [hunk 1]
```diff
@@ -582,15 +580,14 @@ abstract class BCodeIdiomatic extends BCodeGlue {
...
-    final def isInstance(tk: BType) {
-      jmethod.visitTypeInsn(Opcodes.INSTANCEOF, tk.getInternalName)
+    final def isInstance(tk: RefBType) {
+      jmethod.visitTypeInsn(Opcodes.INSTANCEOF, tk.classOrArrayType)
     }
 
     // can-multi-thread
-    final def checkCast(tk: BType) {
-      assert(tk.isRefOrArrayType, s"checkcast on primitive type: $tk")
+    final def checkCast(tk: RefBType) {
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala]: procedure syntax was deprecated, btw

File: src/compiler/scala/tools/nsc/transform/Mixin.scala [hunk 2]
```diff
@@ -391,7 +391,7 @@ abstract class Mixin extends InfoTransform with ast.TreeDSL {
           else {
             sourceModule setPos sym.pos
             if (sourceModule.flags != MODULE) {
-              log("!!! Directly setting sourceModule flags from %s to MODULE".format(sourceModule.flagString))
+              log("!!! Directly setting sourceModule flags for $sourceModule from %s to MODULE".format(sourceModule.flagString))
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/Mixin.scala]: interpolator refactoring interrupted half

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala (hunk 1 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala]: Technically, the past tense only applies in the frame of reference of the terminator that we sent forward in time to deprecate this in the `-Xfuture`. Thoughtfully, the future folk sent a T-1000, currently assuming the form of @samuelgruetter, back to our time to create [srewrite](https://github.com/samuelgruetter/srewrite).

Comment (som-snytt) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala]: The real challenge is to go back in git history without triggering Armageddon.

File: src/compiler/scala/tools/nsc/transform/Mixin.scala (hunk 2 above)
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/transform/Mixin.scala]: This must be under `-Xlint:no-missing-interpolator`.

**Comment by gkossakowski** [comment]
LGTM

Good work!

**Comment by lrytz** [comment]
addressed feedback: introduced `asClassBType` and friends (https://github.com/lrytz/scala/commit/fcc3842d602114b280e0d9e3dc3adeb5f2f027e7#commitcomment-6416102), fixed a half-hearted string interpolation, removed some procedure syntax.

`-Ybackend:GenBCode` build running here: https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/1220/console

**Comment by lrytz** [comment]
PLS REBUILD/pr-scala@4b800a1

**Comment by lrytz** [comment]
PLS REBUILD/pr-scala@3914e2b

**Comment by lrytz** [comment]
PLS REBUILD/pr-scala@ec3bcc2

**Comment by lrytz** [comment]
NOLITTER!

**Comment by som-snytt** [comment]
I've never seen kitteh apologizz for her bad behavior.  Is she feeling OK?

**Comment by adriaanm** [comment]
She's feeling a bit vulnerable these days. Not insecure, though.

**Pull request merged**